# 💻 Week 02 Lab – Working with Weather Data

**LSE DS105W – Data for Data Science (2024/25)**

<div style="font-family: system-ui; padding: 20px 30px 20px 20px; background-color: #FFFFFF; border-left: 8px solid #47315E; border-radius: 8px; box-shadow: 0 4px 12px rgba(0, 0, 0, 0.1);max-width:600px">

**Today's Lab Session**
- 📅 Friday, 31 January 2025
- ⏰ Check your class group time in [Course Info](https://moodle.lse.ac.uk/mod/page/view.php?id=1559985)
- 📍 KSW.2.10

🥅 **Learning Goals**

1. Extract and format structured data from a Python dictionary. 
2. Use the `requests` package to collect data dynamically from OpenMeteo. 
3. Reuse your own code by writing functions. 
4. Extend your understanding of data organisation and Markdown tables.

**Prerequisites**  

You should be comfortable with the following concepts before starting this lab:

- ✅ Navigating a Jupyter Notebook and creating new cells.
- ✅ Using `print()` to curate and format textual output.
- ✅ Basic Markdown formatting (tables, lists, etc.).
- ✅ Working with lists and loops (even if it takes you a while to figure them out).
- ✅ Accessing keys in a dictionary. 

</div>

## Part I: Working with a Fixed Dictionary

In [1]:
weather_data = {
    "hourly": {
        "time": ["2025-01-28T00:00", "2025-01-28T01:00", "2025-01-28T02:00"],
        "rain": [0.0, 0.1, 0.2],
        "temperature_2m": [8, 7, 8]
    }
}

In [ ]:
time = weather_data["hourly"]["time"]
rain = weather_data["hourly"]["rain"]
temperature_2m = weather_data["hourly"]["temperature_2m"]
print(time)
print(rain)
print(temperature_2m)

for i in zip(weather_data["hourly"]):
    print(type(i))
length(time)

['2025-01-28T00:00', '2025-01-28T01:00', '2025-01-28T02:00']
[0.0, 0.1, 0.2]
[8, 7, 8]
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>


In [30]:
for i in zip(time,rain,temperature_2m):
    print(f"| {i[0]} | {i[1]} | {i[2]} |")

| 2025-01-28T00:00 | 0.0 | 8 |
| 2025-01-28T01:00 | 0.1 | 7 |
| 2025-01-28T02:00 | 0.2 | 8 |


| Time | Temp (ºC) | Rain (mm)|
|-|-|-|
| 2025-01-28T00:00 | 0.0 | 8 |
| 2025-01-28T01:00 | 0.1 | 7 |
| 2025-01-28T02:00 | 0.2 | 8 |

In [100]:
import requests
url = "https://api.open-meteo.com/v1/forecast?latitude=51.5085&longitude=-0.1257&hourly=temperature_2m&timezone=Europe%2FLondon"
response = requests.get(url)

In [101]:
# was the request successful?
print(response.status_code) # if 200, successful; if 404, url not found

200


In [102]:
print(response.text)
weather_data = response.json()

{"latitude":51.5,"longitude":-0.120000124,"generationtime_ms":0.034928321838378906,"utc_offset_seconds":0,"timezone":"Europe/London","timezone_abbreviation":"GMT","elevation":23.0,"hourly_units":{"time":"iso8601","temperature_2m":"°C"},"hourly":{"time":["2025-01-31T00:00","2025-01-31T01:00","2025-01-31T02:00","2025-01-31T03:00","2025-01-31T04:00","2025-01-31T05:00","2025-01-31T06:00","2025-01-31T07:00","2025-01-31T08:00","2025-01-31T09:00","2025-01-31T10:00","2025-01-31T11:00","2025-01-31T12:00","2025-01-31T13:00","2025-01-31T14:00","2025-01-31T15:00","2025-01-31T16:00","2025-01-31T17:00","2025-01-31T18:00","2025-01-31T19:00","2025-01-31T20:00","2025-01-31T21:00","2025-01-31T22:00","2025-01-31T23:00","2025-02-01T00:00","2025-02-01T01:00","2025-02-01T02:00","2025-02-01T03:00","2025-02-01T04:00","2025-02-01T05:00","2025-02-01T06:00","2025-02-01T07:00","2025-02-01T08:00","2025-02-01T09:00","2025-02-01T10:00","2025-02-01T11:00","2025-02-01T12:00","2025-02-01T13:00","2025-02-01T14:00","2025

In [103]:
for i in weather_data["hourly"]:
    print(i)

time
temperature_2m


In [104]:
def get_weather_variables(weather_data):
    times = weather_data["hourly"]["time"]
    temps = weather_data["hourly"]["temperature_2m"]
    return times, temps
hourly_weather = get_weather_variables(weather_data)

In [108]:
url = "https://api.open-meteo.com/v1/forecast?latitude=51.5085&longitude=-0.1257&hourly=temperature_2m&hourly=rain&timezone=Europe%2FLondon"
response = requests.get(url)
weather_data = response.json()

In [131]:
def print_weather_table(weather_data):
    times = weather_data["hourly"]["time"]
    temps = weather_data["hourly"]["temperature_2m"]
    rain = weather_data["hourly"]["rain"]
    rows = []
    for i in zip(times,temps,rain):
        rows.append(f"| {i[0]} | {i[1]} | {i[2]} |")
    return f"| Time | Temp (ºC) | Rain (mm)|\n|-|-|-|\n{rows[0]}\n{rows[1]}"
print_weather_table(weather_data)

'| Time | Temp (ºC) | Rain (mm)|\n|-|-|-|\n| 2025-01-31T00:00 | 2.5 | 0.0 |\n| 2025-01-31T01:00 | 2.6 | 0.0 |'

| Time | Temp (ºC) | Rain (mm)|
|-|-|-|
| 2025-01-31T00:00 | 0.0 | 2.5 |

| Time | Temp (ºC) | Rain (mm)|\n|-|-|-|\n| 2025-01-31T00:00 | 2.5 | 0.0 |\n| 2025-01-31T01:00 | 2.6 | 0.0 |